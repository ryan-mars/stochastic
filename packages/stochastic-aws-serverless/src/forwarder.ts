import { PublishCommand, SNSClient } from "@aws-sdk/client-sns";
import { unmarshall } from "@aws-sdk/util-dynamodb";
import { DynamoDBStreamHandler } from "aws-lambda";
import * as env from "env-var";

const sns = new SNSClient({});

export const handler: DynamoDBStreamHandler = async (event) => {
  const EVENT_STREAM_TOPIC_ARN: string = env.get("EVENT_STREAM_TOPIC_ARN").required().asString();
  console.log(JSON.stringify(event, null, 2));
  for (const record of event.Records) {
    try {
      if (record.eventName !== "INSERT") {
        console.log(`Skipping ${record.eventName}`);
        continue
      }

      if (!record.dynamodb?.NewImage) {
        throw new Error(`Missing NewImage item in: ${JSON.stringify(record, null, 2)}`);
      }
      const item = unmarshall(record.dynamodb.NewImage as any);
      const { SequenceNumber } = record.dynamodb;
      const data = await sns.send(
        new PublishCommand({
          Message: JSON.stringify({ ...item, SequenceNumber }),
          TopicArn: EVENT_STREAM_TOPIC_ARN,
          MessageAttributes: {
            event_type: {
              DataType: "String",
              StringValue: item.type ?? "Unspecified",
            },
            version: {
              // Event type version
              DataType: "String",
              StringValue: item.version ?? "0",
            },
          },
        }),
      );
      console.log({
        DynamoDBSequenceNumber: SequenceNumber,
        SNSMessageId: data.MessageId,
      });
    } catch (err) {
      console.error(err, err.stack);
      throw new Error(err);
    }
  }
};
