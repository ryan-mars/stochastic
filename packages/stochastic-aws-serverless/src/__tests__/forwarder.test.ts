const { PublishCommand } = jest.requireActual("@aws-sdk/client-sns")
const mockSend = jest.fn().mockResolvedValue({ MessageId: "165545c9-2a5c-472c-8df2-7ff2be2b3b1b" })

jest.mock("@aws-sdk/client-sns", () => {
  return {
    SNSClient: jest.fn().mockImplementation(() => {
      return { send: mockSend }
    }),
    PublishCommand,
  }
})

import dynamodbStreamEvent from "./__fixtures__/dynamodb-stream-event.json"
import forwarderSNSMessage from "./__fixtures__/forwarder-sns-message.json"
import { handler } from "../forwarder"
import { Context, DynamoDBStreamEvent } from "aws-lambda"
import { SNSClient } from "@aws-sdk/client-sns"

describe("event forwarder", () => {
  globalThis.process.env["EVENT_STREAM_TOPIC_ARN"] = "my:sns:arn"

  it("forward the event to SNS", async () => {
    await handler(dynamodbStreamEvent as any, {} as Context, () => {})
    expect(mockSend.mock.calls[0][0].input).toMatchSnapshot()
    expect(mockSend.mock.calls[0][0].input.Message).toMatch(JSON.stringify(forwarderSNSMessage))
  })
})
