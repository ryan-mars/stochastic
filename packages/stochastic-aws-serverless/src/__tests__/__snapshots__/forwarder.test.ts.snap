// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event forwarder forward the event to SNS 1`] = `
Object {
  "Message": "{\\"payload\\":{\\"flightNo\\":\\"PA123\\",\\"aircraftType\\":\\"B787-9\\",\\"__typename\\":\\"FlightCreated\\",\\"origin\\":\\"SFO\\",\\"destination\\":\\"MIA\\",\\"days\\":{}},\\"sk\\":\\"EVENT#1sS1oCIxZgbic90cQRz5rYUcDwY\\",\\"id\\":\\"1sS1oCIxZgbic90cQRz5rYUcDwY\\",\\"pk\\":\\"FlightSchedule#PA123\\",\\"source\\":\\"FlightSchedule\\",\\"source_id\\":\\"PA123\\",\\"time\\":\\"2021-05-12T20:29:08.906Z\\",\\"type\\":\\"FlightCreated\\",\\"SequenceNumber\\":\\"100000000009009450812\\"}",
  "MessageAttributes": Object {
    "event_type": Object {
      "DataType": "String",
      "StringValue": "FlightCreated",
    },
    "version": Object {
      "DataType": "String",
      "StringValue": "0",
    },
  },
  "TopicArn": "my:sns:arn",
}
`;
